src/vs/workbench/contrib/chat/browser/contrib/chatDynamicVariables.ts

```diff
@@ -60,60 +67,120 @@ export class ChatDynamicVariableModel extends Disposable implements IChatWidgetC
 							this.widget.refreshParsedInput();
 						}
 						return null;
-					} else if (Range.compareRangesUsingStarts(ref.range, c.range) > 0) {
+					}
+
+					if (Range.compareRangesUsingStarts(ref.range, c.range) > 0) {
 						const delta = c.text.length - c.rangeLength;
-						return {
-							...ref,
-							range: {
-								startLineNumber: ref.range.startLineNumber,
-								startColumn: ref.range.startColumn + delta,
-								endLineNumber: ref.range.endLineNumber,
-								endColumn: ref.range.endColumn + delta
-							}
+						ref.range = {
+							startLineNumber: ref.range.startLineNumber,
+							startColumn: ref.range.startColumn + delta,
+							endLineNumber: ref.range.endLineNumber,
+							endColumn: ref.range.endColumn + delta,
 						};
+
+						return ref;
 					}
 
 					return ref;
 				}));
 			});
 
-			this.updateDecorations();
+			this.updateVariableDecorations();
 		}));
 	}
 
-	getInputState(): any {
-		return this.variables;
-	}
-
 	setInputState(s: any): void {
 		if (!Array.isArray(s)) {
 			s = [];
 		}
 
 		this._variables = s;
-		this.updateDecorations();
+		this.updateVariableDecorations();
 	}
 
-	addReference(ref: IDynamicVariable): void {
-		this._variables.push(ref);
-		this.updateDecorations();
+	public addReference(ref: IDynamicVariable): void {
+		const variable = this._register(
+			this.instantiationService.createInstance(ChatDynamicVariable, ref),
+		);
+
+		this._variables.push(variable);
+		this.updateVariableDecorations();
 		this.widget.refreshParsedInput();
+
+		// if the `prompt snippets` feature is enabled, start resolving
+		// nested file references immediatelly and subscribe to updates
+		if (variable.isPromptSnippetFile) {
+			// subscribe to variable changes
+			this._register(variable.onUpdate(() => {
```

Comment:
Seems like the wrong lifecycle, these listeners will never be disposed as long as the chat widget exists

Reply:
Added logic to dispose these variables when they are removed from the `_variables` list.

Reply:
> We have to add to disposables here anyway

Why?

Reply:
> Why?

I've written this and then had the second thought - you are right we don't have to since we now dispose the variables explicitly 🙃

Reply:
Changed `onUpdate` to not return a disposable at all.